File: orchestrator/internal/wire/secrets.go
// Package wire reproduces the freehold-core secret + Nostr wire formats
// (core/src/secrets.rs, core/src/nip98.rs, core/src/audit.rs,
// core/src/memory.rs) byte-exactly in Go.
package wire

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// SECRETS_FILE is the on-disk name of a shipped secret package.
const SECRETS_FILE = "secrets.json"

// TargetMeta is per-target connector metadata shipped alongside ciphertext.
// Mirrors core/src/secrets.rs::TargetMeta.
type TargetMeta struct {
	Kind    string `json:"kind"`
	Address string `json:"address"`
	Secret  string `json:"secret"`
}

// SecretPackage is the control plane's shipped package: secret name -> sealed
// ciphertext (hex). Mirrors core/src/secrets.rs::SecretPackage.
//
// Field order and the always-present targets/grants (serde default) matter;
// encoding/json sorts map keys, so a plain struct + maps yields byte-identical
// output to serde's BTreeMap. Empty grants must serialize as [] (not null),
// so the JSON marshaling path below initializes them explicitly.
type SecretPackage struct {
	Secrets map[string]string     `json:"secrets"`
	Targets map[string]TargetMeta `json:"targets"`
	Grants  []string              `json:"grants"`
}

// jsonPackage orders fields secrets -> targets -> grants (serde struct order).
type jsonPackage struct {
	Secrets map[string]string     `json:"secrets"`
	Targets map[string]TargetMeta `json:"targets"`
	Grants  []string              `json:"grants"`
}

// New builds a SecretPackage with non-nil maps/slice so JSON emission matches
// serde (targets: {}, grants: [] even when empty).
func New(secrets map[string]string, targets map[string]TargetMeta, grants []string) *SecretPackage {
	if secrets == nil {
		secrets = map[string]string{}
	}
	if targets == nil {
		targets = map[string]TargetMeta{}
	}
	if grants == nil {
		grants = []string{}
	}
	return &SecretPackage{Secrets: secrets, Targets: targets, Grants: grants}
}

// MarshalJSON emits serde-compatible JSON (sorted keys, empty containers).
func (p *SecretPackage) MarshalJSON() ([]byte, error) {
	jp := jsonPackage{
		Secrets: p.Secrets,
		Targets: p.Targets,
		Grants:  p.Grants,
	}
	if jp.Secrets == nil {
		jp.Secrets = map[string]string{}
	}
	if jp.Targets == nil {
		jp.Targets = map[string]TargetMeta{}
	}
	if jp.Grants == nil {
		jp.Grants = []string{}
	}
	return json.Marshal(jp)
}

// WriteToDir writes the package atomically (0600) into dir (created 0700).
// Mirrors core::secrets::SecretPackage::write_to_dir.
func (p *SecretPackage) WriteToDir(dir string) error {
	if err := ensurePrivateDir(dir); err != nil {
		return err
	}
	data, err := json.Marshal(p)
	if err != nil {
		return fmt.Errorf("marshal secret package: %w", err)
	}
	return write0600Atomic(filepath.Join(dir, SECRETS_FILE), data)
}

// Load reads a previously shipped package from dir. Mirrors
// core::secrets::SecretPackage::load.
func Load(dir string) (*SecretPackage, error) {
	raw, err := os.ReadFile(filepath.Join(dir, SECRETS_FILE))
	if err != nil {
		return nil, err
	}
	p := &SecretPackage{}
	if err := json.Unmarshal(raw, p); err != nil {
		return nil, err
	}
	return p, nil
}

// Bytes returns the serde-compatible JSON bytes for byte-identity checks.
func (p *SecretPackage) Bytes() ([]byte, error) {
	var buf bytes.Buffer
	data, err := json.Marshal(p)
	if err != nil {
		return nil, err
	}
	buf.Write(data)
	return buf.Bytes(), nil
}

File: control-plane/core/src/secrets.rs
//! The runner's on-disk secret config: name -> sealed ciphertext (hex).
//!
//! WRITTEN by the control plane at provision/rotate; READ by the runner
//! (Phase A4+), which opens each ciphertext with its injected encryption key.
//! Plaintext never sits in this file, in the CP's records, or in agent
//! context — the runner references secrets BY NAME only.

use std::collections::BTreeMap;
use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::futil;

pub const SECRETS_FILE: &str = "secrets.json";

/// Non-secret per-target metadata the CP ships alongside the ciphertext so
/// the runner can enumerate what it reaches and how to connect.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TargetMeta {
    pub kind: String,
    /// Connector-specific address (e.g. `user@host:port` for ssh).
    pub address: String,
    /// Which entry in `secrets` holds this target's credential.
    pub secret: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct SecretPackage {
    /// secret name -> sealed-box ciphertext (hex)
    pub secrets: BTreeMap<String, String>,
    /// target name -> connector metadata (local is implicit; kind: ssh, …)
    #[serde(default)]
    pub targets: BTreeMap<String, TargetMeta>,
    /// AGENT pubkeys (Nostr x-only, hex) allowed to call this runner.
    /// Enforced locally at the MCP boundary; ports to relay membership in
    /// Chunk 2. Empty = nobody may call (fail closed).
    #[serde(default)]
    pub grants: Vec<String>,
}

impl SecretPackage {
    /// Atomic 0600 write into `dir` (creates the dir 0700).
    pub fn write_to_dir(&self, dir: &Path) -> std::io::Result<()> {
        futil::ensure_private_dir(dir)?;
        let json = serde_json::to_vec_pretty(self)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
        futil::write_0600_atomic(&dir.join(SECRETS_FILE), &json)
    }

    /// Load a package previously shipped to `dir` for the runner.
    pub fn load(dir: &Path) -> std::io::Result<Self> {
        let raw = std::fs::read_to_string(dir.join(SECRETS_FILE))?;
        serde_json::from_str(&raw)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn package_roundtrip() {
        let dir = tempfile::tempdir().unwrap();
        let pkg = SecretPackage {
            secrets: BTreeMap::from([("vultr".into(), "ciphertext-hex".into())]),
            targets: BTreeMap::from([(
                "vultr".into(),
                TargetMeta {
                    kind: "vultr".into(),
                    address: "api.vultr.com".into(),
                    secret: "vultr".into(),
                },
            )]),
            grants: vec!["agent-pubkey".into()],
        };
        pkg.write_to_dir(dir.path()).unwrap();
        let loaded = SecretPackage::load(dir.path()).unwrap();
        assert_eq!(loaded, pkg);
    }

    #[cfg(unix)]
    #[test]
    fn package_is_0600() {
        use std::os::unix::fs::PermissionsExt;
        let dir = tempfile::tempdir().unwrap();
        SecretPackage::default().write_to_dir(dir.path()).unwrap();
        let mode = std::fs::metadata(dir.path().join(SECRETS_FILE))
            .unwrap()
            .permissions()
            .mode();
        assert_eq!(
            mode & 0o077,
            0,
            "secrets.json must not be group/other readable"
        );
    }
}
